### PR TITLE
feat: indexed access types

### PR DESCRIPTION
**Pull request by harryparkdotio**
allow indexed access type syntax

```ts
interface Company {
  id: string;
}

@request
function request(
  @pathParams
  pathParams: {
    companyId: Company['id']
  }
)
```

---

### note

I haven't implemented nested indexed access type support yet

i.e.

```ts
interface Company {
  data: {
    id: string;
  }
}

@request
function request(
  @pathParams
  pathParams: {
    companyId: Company['data']['id']
  }
)
```

### CODE REVIEW

**Comment by lfportal** [comment]
hey @harryparkdotio, could you create an issue to discuss/track this feature?

**Comment by harryparkdotio** [comment]
sure, specifically to track nested indexed access types, or this feature as a whole?

File: lib/src/parsers/utilities/type-parser.ts [hunk 1]
```diff
@@ -217,6 +220,53 @@ function parseUnionType(typeNode: UnionTypeNode): DataType {
...
+    throw new Error(
+      "indexed access type error: nested indexed access type unsupported"
+    );
+  }
+
+  if (!TypeGuards.isTypeReferenceNode(object)) {
+    throw new Error("indexed access type error");
+  }
+
+  const declaration = getTargetDeclarationFromTypeReference(object);
```
Comment (medric) [inline lib/src/parsers/utilities/type-parser.ts]: AFAIK, `getTargetDeclarationFromTypeReference` would always return a `declaration` or throw an error. @lfportal @fwouts can you confirm the logic here please ?

File: lib/src/parsers/utilities/type-parser.ts [hunk 2]
```diff
@@ -217,6 +220,53 @@ function parseUnionType(typeNode: UnionTypeNode): DataType {
...
+
+  if (TypeGuards.isIndexedAccessTypeNode(object)) {
+    // FIXME: add support for nested indexed access types
+    throw new Error(
+      "indexed access type error: nested indexed access type unsupported"
+    );
+  }
+
+  if (!TypeGuards.isTypeReferenceNode(object)) {
+    throw new Error("indexed access type error");
```
Comment (medric) [inline lib/src/parsers/utilities/type-parser.ts]: As per, the previous `throw` statement. Can you add a bit more context to the exception(s) being thrown, as in more helpful messages.

File: lib/src/parsers/utilities/type-parser.ts [hunk 3]
```diff
@@ -217,6 +220,53 @@ function parseUnionType(typeNode: UnionTypeNode): DataType {
...
+ * Parse an indexed access type node.
+ *
+ * @param typeNode index access type node
+ */
+function parseIndexedAccessType(typeNode: IndexedAccessTypeNode): TypeNode {
+  const object = typeNode.getObjectTypeNode();
+  const index = typeNode.getIndexTypeNode();
+
+  if (TypeGuards.isIndexedAccessTypeNode(object)) {
+    // FIXME: add support for nested indexed access types
```
Comment (lfportal) [inline lib/src/parsers/utilities/type-parser.ts]: Can we action this fix now?

File: lib/src/parsers/utilities/type-parser.ts (hunk 1 above)
Comment (lfportal) [reply lib/src/parsers/utilities/type-parser.ts]: yes this function will throw an error

**Review by medric — APPROVED** [review]
LGTM, thanks. Just address the lint errors please.

**Pull request merged**